packages/template-retail-react-app/app/utils/shipment-utils.js

```diff
@@ -0,0 +1,246 @@
+ */
+export const findUnusedDeliveryShipment = (basket, usedShipmentIds = [], isPickupMethod) => {
+    if (!basket?.shipments) return null
+
+    return (
+        basket.shipments.find(
+            (shipment) =>
+                !isPickupMethod(shipment.shippingMethod) &&
+                !usedShipmentIds.includes(shipment.shipmentId)
+        ) || null
+    )
+}
+
+/**
+ * Compares two addresses to determine if they are the same
+ * @param {Object} address1 - First address object
+ * @param {Object} address2 - Second address object
+ * @returns {boolean} True if addresses match
+ */
+export const areAddressesEqual = (address1, address2) => {
+    // Normalize falsey values (null, undefined, empty string)
+    const normalize = (value) => (!value ? '' : value)
+
+    return (
+        normalize(address1?.firstName) === normalize(address2?.firstName) &&
+        normalize(address1?.lastName) === normalize(address2?.lastName) &&
+        normalize(address1?.address1) === normalize(address2?.address1) &&
+        normalize(address1?.city) === normalize(address2?.city) &&
+        normalize(address1?.stateCode) === normalize(address2?.stateCode) &&
+        normalize(address1?.postalCode) === normalize(address2?.postalCode) &&
+        normalize(address1?.countryCode) === normalize(address2?.countryCode)
+    )
+}
+
+/**
+ * Extracts only valid OrderAddress fields from an address object
+ * @param {Object} address - The address object (may contain extra fields from customer address)
+ * @returns {Object} Clean address object with only OrderAddress fields
+ */
+export const cleanAddressForOrder = (address) => {
+    if (!address) return null
+
+    return {
+        address1: address.address1,
+        city: address.city,
+        countryCode: address.countryCode,
+        firstName: address.firstName,
+        lastName: address.lastName,
+        phone: address.phone,
+        postalCode: address.postalCode,
+        stateCode: address.stateCode
+    }
+}
+
+/**
+ * Finds the first existing delivery shipment with matching address
+ * @param {Object} basket - The basket object
+ * @param {Object} address - The address to match
+ * @param {Function} isPickupMethod - Function to check if a shipping method is pickup
+ * @returns {string|null} The shipment ID with matching address or null if not found
+ */
+export const findDeliveryShipmentWithSameAddress = (basket, address, isPickupMethod) => {
+    if (!basket?.shipments || !address) return null
+
+    const foundShipment = basket.shipments.find((shipment) => {
+        // Must be a delivery shipment (not pickup)
+        if (isPickupMethod(shipment.shippingMethod)) {
+            return false
+        }
+
+        // Check if shipment has a shipping address that matches
+        return shipment.shippingAddress && areAddressesEqual(shipment.shippingAddress, address)
+    })
+    return foundShipment?.shipmentId || null
```

Comment:
Either use `findDeliveryShipmentIdWithSameAddress` or return shipment object. `shipment` indicates an object but we are returning a string here

Reply:
return shipment object gives more flexibility imo